### DISCUSSION TITLE
JPM uncertainty treatment

### DISCUSSION THREAD

**Discussion by fgarcia0128** [coastal-hazards-system/StormSim-Library#11] [Announcements]
Background:
The latest version of StormSim is composed of 7 main steps:
- Step 1: Set-up Environment
- Step 2: Import, Process & Format CHS Data
- Step 3: Create Storm Forcing
- Step 4: Create Structure Geometry
- Step 5: Apply Uncertainty To Project Forcing -> Latest PCHA method, U_a, U_r (combined), SWL,Hm0, Tp
- Step 6: Apply Project Forcing  Adjustments  -> Adds SLR, Random Tide, Depth Limitation
- Step 7: Compute Project Responses 

These 7 steps support both Response Base (RB) & Life Cycle Base (LCS) workflows. Currently, project responses supported are:
- LCS-> S (damage progression), Reliability & Probability of Failure (CSR Peaks)
- RB1 & RB3 -> SST/JPM -> q, R2p, R2p+SWL, Dn50, Dn50_LCBW, P1, SWL, Hm0, Tp 

The main take away from this is that uncertainty is a stand alone step outside any computational routine. This provides transparency and better control/implementation of how we apply uncertainty across our workflows.

Problem:
This setup suggest we would handle forcing uncertainty outside SST/JPM. This however, results in HCs for JPM with all CL's on top of one another (due JPM not applying uncertainty internally). SST already has a flag to turn off internal uncertainty and compute CL's as percentiles.

Question:
Could we implement something similar for JPM ?

Bottom figure is with uncertainty being applied by SST/JPM internally.

@NCNC-628 @jmelby @LukeAucoin17 

![Midbay_v1_James_Island_StormSim_Maxima_Project_Forcing_and_Hazard_Curve_Comparison](https://user-images.githubusercontent.com/51959561/224115459-c78113ef-2fbb-4a0c-bded-49330e2b3aa9.png)
![Midbay_v1_James_Island_StormSim_Maxima_Project_Forcing_and_Hazard_Curve_Comparison](https://user-images.githubusercontent.com/51959561/224115513-12d1168e-53b4-43bc-a1b6-d46934efbd2f.png)

**Comment by LukeAucoin17** [comment]


**Reply by fgarcia0128** [reply]
I think what you are suggesting is to grab the best estimates results (with uncertainty applied in step 5) and then compute the CLs from that as percentiles in Step 7. (Right after responses come out of JPM).

**Comment by jmelby** [comment]
Let me see if I can summarize where we are.  Bias and uncertainty are handled differently in StormSim than in the typical regional PCHA.  Bias correction is applied to CHS regional study data in call_chs_data_formater --> call_chs_storm_quality_check.  For project studies with separate hydrodynamic modeling, there is no software to format files and add bias correction.  This is okay because there are no standards for hydro output so they have a lot of variability but CHG can assist each project to develop routines as needed.
There are several workflows for uncertainty:
1.	SWL and Hm0 TC: expand input arrays using 444 replicates in rb_forcing_formater.m.  Replicates are inputs to JPM integration code where U_a and U_r are applied as in Regional PCHA
2.	SWL and Hm0 XC:  Bare POT sample (size nStorm) is input to SST code where U_a and U_r are applied as in Regional PCHA using 1000 bootstrap random samples.
3.	Structure Responses TC:  444 replicates of wave and water level parameters are generated. Wave and water level uncertainty (U_a and U_r) should be applied to these and then structure responses computed.  These 444*nstorms structure responses are inputs to JPM integration code where structure equation U_r is applied.
4.	Structure Responses XC:  20 replicates are generated and wave and water level uncertainty (U_a and U_r) applied.  Structure responses are computed for these 20 * nStorm forcing replicates.  These are inputs to JPM integration code where structure equation U_r is applied.
We still need to do convergence analysis to determine the best values for the 444 and 20 replicates.  These are arbitrary now.  We also need to do convergence tests for the 1000 bootstrap nSim values.  I am guessing this is way overkill.

**Comment by jmelby** [comment]
One problem that has not been addressed... in the old PCHA procedure, we added part of the uncertainty to the best estimate in order to add many random realizations in order to produce a smooth BE HC.  In the new PCHA method, this is not needed if the surrogate model is used to produce many more realizations.  However, for StormSim, we dont have the surrogate implemented.  The result is that no uncertainty is applied to the BE and so the BE HC can be coarse (jagged).  We cannot require project studies to produce a surrogate so that is not an option for StormSim in the near term.  We could apply a small amount of uncertainty to the BE with many realizations to get a smooth curve.  Something to discuss.
